SPEC.md

```diff
@@ -0,0 +1,225 @@
+
+<p align="center">
+  <br>
+  <img src="images/context.png" alt="Fetcher Stack"/>
+</p>
+
+## API
+
+The fetcher interface contains a single method:
+
+```js
+interface ApolloFetcher{
+  request(Operation) : Observable<FetchResult>
+}
+```
+
+An Operation contains the AST for a particular query, variables and associated context:
+
+```js
+Operation {
+  query: DocumentNode,
+  operationName: string,
+  variables: object,
+  context: object,
+}
+```
+
+An `Observable` is returned from the `request` function and controls the life-cycle of a GraphQL request:
+
+```js
+Observable{
+  //Causes the underlying producer to start
+  subscribe(Subscriber<FetchResult>) : UnsubscribeHandler
+  subscribe(
+    next(FetcheResult) => void,
+    error(any) => void,
+    complete() => void) : UnsubscribeHandler
+  status(): any
+}
+```
+
+A Subscriber is passed to the `subscribe` function
+
+```js
+Subscriber<T>{
+  next(T),
+  error(any),
+  complete(),
+}
+```
+
+UnsubscribeHandler is returned from a `subscribe` function in the Observable
+
+```js
+type UnscubscribeHandler = () => void
+```
+
+FetchResult is passed to the `next` function in the Observable
+
+```js
+FetchResult{
+  data?: object,
+  errors?: object[],
+  extensions?: any,
+  context?:object,
+}
+```
+
+* Should explain what might go into context.
+
+## More examples
+
+Here is a list of planned fetchers with selected diagrams:
+
+Base
+
+* Websocket: subscriptions support
+* Batching-\*: support for batching operations with a transport
+
+<p align="center">
+  <br>
+  <img src="images/batch-fetcher.png" alt="Batch Fetcher"/>
+</p>
+
+Intermediate
+
+* Polling: repeats requests on a specified interval
+
+<p align="center">
+  <br>
+  <img src="images/polling-fetcher.png" alt="Polling Fetcher"/>
+</p>
+
+* Caching: returns data if result in the cache and stores data in cache on response
+* Compose: combines a list of fetchers (would require additional semantics around constructor argument order)
+* Dedup: saves query signatures that are currently on the wire and returns the result for all of those queries
+* NPlusOne: adds query based on predicted access
+* Retry: error callback causes the request to retry
+* Mock: returns fake data for all or part of a request ← implement with BaseFetcher (eg. executes `graphql`)
+
+Forks:
+
+* Split: split operations between fetchers depending on a function passed in
+* Hybrid: uses split-fetcher to fill query and mutations with http and subscriptions with websockets
+
+<p align="center">
+  <br>
+  <img src="images/hybrid-fetcher.png" alt="Hybrid Fetcher"/>
+</p>
+
+Adapters:
+
+* Wrapper => returns a promise : not a fetcher
+
+<p align="center">
+  <br>
+  <img src="images/fetcher-as-promise.png" alt="Fetcher Promise Wrapper"/>
+</p>
+
+## Open questions, discussion
```

Comment:
This section is a bit unpolished. Maybe you can clean it up by splitting it into subsections for each discussion topic, and writing a sentence or two about what should be discussed. For example

```
### How to pass context to the `next` callback?
Should we pass the context from the response (for example relevant HTTP headers or status code) to the `next` callback of the observable like this?: `next({ data, errors, extensions, context }){ }` or like this: `{ result: { data, errors, extensions }, context }`.

```

etc.